crates/history/src/history.rs

```diff
@@ -0,0 +1,102 @@
+use serde::Serialize;
+
+use crate::error::HistoryResult;
+use crate::listener::HistoryListener;
+use crate::location::Location;
+
+/// A trait to provide [`History`] access.
+pub trait History: Clone + PartialEq {
+    type Location: Location<History = Self> + 'static;
+
+    /// Returns the number of elements in [`History`].
+    fn len(&self) -> usize;
+
+    /// Returns true if the current [`History`] is empty.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Moves back 1 page in [`History`].
+    fn back(&self) {
+        self.go(-1);
+    }
+
+    /// Moves forward 1 page in [`History`].
+    fn forward(&self) {
+        self.go(1);
+    }
+
+    /// Loads a specific page in [`History`] with a `delta` relative to current page.
+    ///
+    /// See: <https://developer.mozilla.org/en-US/docs/Web/API/History/go>
+    fn go(&self, delta: isize);
+
+    /// Pushes a route entry with [`None`] being the state.
+    fn push(&self, route: impl Into<String>);
```

Comment:
```suggestion
    fn push(&self, route: impl AsRef<str>);
```

Avoid the owned `String` allocation

Reply:
This is actually deliberate.

This is due to routes(`Routable`) for `yew-router` is usually declared as an enum, and can implement `Into<String>` easily, but cannot implement `AsRef<str>` easily.

I guess maybe we can do `Into<Cow<'_, str>>` instead?

Reply:
> I guess maybe we can do `Into<Cow<'_, str>>` instead?

That would be good. That way we can avoid the allocation when possible